src/main/cpp/src/hive_hash.cu

```diff
@@ -486,16 +526,70 @@ std::unique_ptr<cudf::column> hive_hash(cudf::table_view const& input,
+        root_col.type().id() == cudf::type_id::STRUCT) {
+      // Construct the `flattened_column_views` by level order traversal
+      nested_column_map.push_back(flattened_column_views.size());
+      flattened_column_views.push_back(root_col);
+      // flattened_column_views[idx] is the next column to process
+      for (auto idx = nested_column_map.back();
+           idx < static_cast<cudf::size_type>(flattened_column_views.size());
+           idx++) {
+        auto const col = flattened_column_views[idx];
+        if (col.type().id() == cudf::type_id::LIST) {
+          first_child_index.push_back(flattened_column_views.size());
+          flattened_column_views.push_back(cudf::lists_column_view(col).offsets());
+          flattened_column_views.push_back(cudf::lists_column_view(col).get_sliced_child(stream));
+        } else if (col.type().id() == cudf::type_id::STRUCT) {
+          first_child_index.push_back(flattened_column_views.size());
+          for (auto child_idx = 0; child_idx < col.num_children(); child_idx++) {
+            flattened_column_views.push_back(
+              cudf::structs_column_view(col).get_sliced_child(child_idx, stream));
+          }
+        } else {
+          first_child_index.push_back(-1);
+        }
+      }
+    } else {
+      nested_column_map.push_back(-1);
+    }
+  }
+
+  [[maybe_unused]] auto [device_view_owners, flattened_column_device_views] =
+    cudf::contiguous_copy_column_device_views<cudf::column_device_view>(flattened_column_views,
+                                                                        stream);
+  auto first_child_index_view = cudf::detail::make_device_uvector_async(
+    first_child_index, stream, cudf::get_current_device_resource_ref());
+  auto nested_column_map_view = cudf::detail::make_device_uvector_async(
+    nested_column_map, stream, cudf::get_current_device_resource_ref());
+
   bool const nullable   = has_nested_nulls(input);
   auto const input_view = cudf::table_device_view::create(input, stream);
   auto output_view      = output->mutable_view();
 
   // Compute the hash value for each row
-  thrust::tabulate(rmm::exec_policy(stream),
+  thrust::tabulate(rmm::exec_policy_nosync(stream),
                    output_view.begin<hive_hash_value_t>(),
                    output_view.end<hive_hash_value_t>(),
-                   hive_device_row_hasher<hive_hash_function, bool>(nullable, *input_view));
-
+                   hive_device_row_hasher<hive_hash_function, bool>(nullable,
+                                                                    *input_view,
+                                                                    flattened_column_device_views,
+                                                                    first_child_index_view.data(),
+                                                                    nested_column_map_view.data()));
+
+  // Ensure that the output is ready before returning
```

Comment:
```suggestion
  // Push data from host vectors `first_child_index` and `nested_column_map` to device
  // before they are destroyed.
```

Reply:
It is totally fine to return a column that is not "ready". Indeed, upon returning to the caller, all the kernels have been submitted to device but they may not be executed, leaving the output column still uninitialized. And that is not an issue at all. When the output column is used elsewhere, as long as the same CUDA stream is used to process it, the driver guarantees that the kernels generating its data are always called before the kernels using it. And this is the principle of stream-ordered processing.